packages/SwingSet/src/kernel/kernel.js

```diff
@@ -308,33 +308,35 @@ export default function buildKernel(
     }
   }
 
-  function removeVatManager(vatID, reason) {
+  function removeVatManager(vatID, shouldReject, info) {
+    insistCapData(info);
     const old = ephemeral.vats.get(vatID);
     ephemeral.vats.delete(vatID);
-    const err = reason ? Error(reason) : undefined;
-    old.notifyTermination(err);
+    old.notifyTermination(shouldReject, info);
     return old.manager.shutdown();
   }
 
-  function terminateVat(vatID, reason) {
+  function terminateVat(vatID, shouldReject, info) {
+    insistCapData(info);
     if (kernelKeeper.getVatKeeper(vatID)) {
       const promisesToReject = kernelKeeper.cleanupAfterTerminatedVat(vatID);
-      const err = reason ? makeError(reason) : VAT_TERMINATION_ERROR;
+      const err = VAT_TERMINATION_ERROR;
       for (const kpid of promisesToReject) {
         resolveToError(kpid, err, vatID);
       }
-      removeVatManager(vatID, reason).then(
+      removeVatManager(vatID, shouldReject, info).then(
         () => kdebug(`terminated vat ${vatID}`),
         e => console.error(`problem terminating vat ${vatID}`, e),
       );
     }
   }
 
-  let deliveryProblem;
+  let terminationTrigger;
 
-  function registerDeliveryProblem(vatID, problem, resultKP) {
-    if (!deliveryProblem) {
-      deliveryProblem = { vatID, problem, resultKP };
+  function setTerminationTrigger(vatID, shouldAbortCrank, shouldReject, info) {
+    assert(!(shouldAbortCrank && !shouldReject));
```

Comment:
nit: my brain can't read that without writing out a truth table. Would it be correct to replace it with:

```js
if (shouldAbortCrank) {
  assert(shouldReject);
}
```
?

Reply:
Yeah, I went around with that one a couple of times myself, applying deMorgan's rule and then unapplying it again in a fruitless quest to find which formulation was clearer.  For some reason I was stuck thinking all the logic needed to be inside the assert parameter, which as you point out is not true.